File: 程式設計：資料結構答案檔/C/203.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int value;
    struct TreeNode* left;
    struct TreeNode* right;
};

struct TreeNode* createNode(int value);
struct TreeNode* insertNode(struct TreeNode* root, int value);
void levelOrderTraversal(struct TreeNode* root, int isFirst);
void preorderTraversal(struct TreeNode* root, int isFirst);

struct TreeNode* createNode(int value) {
    struct TreeNode* newNode = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    newNode->value = value;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

struct TreeNode* insertNode(struct TreeNode* root, int value) {
    if (root == NULL) {
        return createNode(value);
    } else {
        if (value < root->value) {
            root->left = insertNode(root->left, value);
        } else {
            root->right = insertNode(root->right, value);
        }
    }
    return root;
}

void preorderTraversal(struct TreeNode* root, int isFirst) {
    if (root == NULL) {
        return;
    }
    if (!isFirst) {
        printf(" ");
    }
    printf("%d", root->value);
    preorderTraversal(root->left, 0);
    preorderTraversal(root->right, 0);
}

void levelOrderTraversal(struct TreeNode* root, int isFirst) {
    if (root == NULL) {
        return;
    }
    struct TreeNode** queue = (struct TreeNode**)malloc(sizeof(struct TreeNode*));
    int front = 0, rear = 0;

    queue[rear++] = root;

    while (front < rear) {
        struct TreeNode* currentNode = queue[front++];
        if (!isFirst) {
            printf(" ");
        }
        printf("%d", currentNode->value);
        isFirst = 0;

        if (currentNode->left != NULL) {
            queue = (struct TreeNode**)realloc(queue, (rear + 1) * sizeof(struct TreeNode*));
            queue[rear++] = currentNode->left;
        }

        if (currentNode->right != NULL) {
            queue = (struct TreeNode**)realloc(queue, (rear + 1) * sizeof(struct TreeNode*));
            queue[rear++] = currentNode->right;
        }
    }
    free(queue);
}

int main() {
    int n, i;
    struct TreeNode* root = NULL;
    scanf("%d", &n);
    int arr[n];
    for (i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }
    for (i = 0; i < n; i++) {
        root = insertNode(root, arr[i]);
    }
    printf("Level-order:");
    levelOrderTraversal(root, 1);
    
    printf("\nPreorder:");
    preorderTraversal(root, 1);

    return 0;
}

File: 程式設計：資料結構答案檔/Go/203.go
package main

import (
	"fmt"
)

type TreeNode struct {
	value int
	left  *TreeNode
	right *TreeNode
}

func createNode(value int) *TreeNode {
	return &TreeNode{value: value}
}

func (root *TreeNode) insertNode(value int) *TreeNode {
	if root == nil {
		return createNode(value)
	}
	if value < root.value {
		root.left = root.left.insertNode(value)
	} else {
		root.right = root.right.insertNode(value)
	}
	return root
}

func preorderTraversal(root *TreeNode, isFirst *bool) {
	if root == nil {
		return
	}
	if !*isFirst {
		fmt.Print(" ")
	}
	fmt.Print(root.value)
	*isFirst = false
	preorderTraversal(root.left, isFirst)
	preorderTraversal(root.right, isFirst)
}

func levelOrderTraversal(root *TreeNode) {
	if root == nil {
		return
	}
	queue := []*TreeNode{root}
	isFirst := true
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		if !isFirst {
			fmt.Print(" ")
		}
		fmt.Print(node.value)
		isFirst = false

		if node.left != nil {
			queue = append(queue, node.left)
		}
		if node.right != nil {
			queue = append(queue, node.right)
		}
	}
}

func main() {
	var n, value int
	fmt.Scan(&n)

	var root *TreeNode
	for i := 0; i < n; i++ {
		fmt.Scan(&value)
		if root == nil {
			root = createNode(value)
		} else {
			root.insertNode(value)
		}
	}

	fmt.Print("Level-order:")
	levelOrderTraversal(root)

	fmt.Print("\nPreorder:")
	isFirst := true
	preorderTraversal(root, &isFirst)
}
